mutable_set/mutable_set.mbt

```diff
@@ -0,0 +1,571 @@
+/// 
+/// ```
+/// println(MutableSet::[1, 2, 3].difference(MutableSet::[4, 5, 1]).to_array())
+/// // output: [3, 2]
+/// ```
+pub fn difference[T : Hash + Eq](
+  self : MutableSet[T],
+  other : MutableSet[T]
+) -> MutableSet[T] {
+  let res = self.copy()
+  other.iter(fn(key : T) -> Unit { res.remove_without_return(key) })
+  res
+}
+
+/// Returns the difference between self and other, but store the result in `into`
+/// 
+/// # Example
+/// 
+/// ```
+/// let into = MutableSet::[]
+/// MutableSet::[1, 2, 3].difference_into(MutableSet::[4, 5, 1], into)
+/// println(into.to_array()) // output: [2, 3]
+/// ```
+pub fn difference_into[T : Hash + Eq](
+  self : MutableSet[T],
+  other : MutableSet[T],
+  into : MutableSet[T]
+) -> Unit {
+  self.difference(other).iter(fn(key : T) -> Unit { into.insert(key) })
+}
+
+// The following are the helper functions used by the internal implementation of MutableSet:
+
+fn Item::new[T](key : T) -> Item[T] {
+  { key, probe_length: 0 }
+}
+
+/// Check whether the threshold is exceeded, and if so, expand the capacity.
+fn check_and_grow[T : Hash + Eq](self : MutableSet[T]) -> Unit {
+  if self.directory.length() * 80 / 100 == self.effective_size {
+    let old = self.copy()
+    let directory = @array.new(
+      self.real_size * 2,
+      fn() -> Option[Item[T]] { None },
+    )
+    self.effective_size = 0
+    self.real_size = 0
+    self.directory = directory
+    old.iter(fn(key : T) -> Unit { self.insert(key) })
+  }
+}
+
+fn insert_item[T : Hash + Eq](self : MutableSet[T], key : T) -> Int {
+  match self.index_of(key) {
+    // When an element already exists, replace it
+    Some(index) => {
+      self.directory[index].unwrap().key = key
+      index
+    }
+    None => {
+      self.check_and_grow()
+      let num_slots = self.directory.length()
+      let slot = key.hash().abs() % num_slots
+      let mut current_probe_length = 0
+      let mut current_item = Item::new(key)
+      let mut index = slot
+      while not(self.directory[index].is_empty()) {
+        let item = self.directory[index].unwrap()
+        if current_probe_length > item.probe_length {
+          current_item.probe_length = current_probe_length
+          let replaced = self.directory[index].unwrap()
+          self.directory[index] = Some(current_item)
+          current_item = replaced
+          current_probe_length = current_item.probe_length
+        }
+        index = (index + 1) % num_slots
+        if index == slot {
+          break
+        }
+        current_probe_length = current_probe_length + 1
+      }
+      current_item.probe_length = current_probe_length
+      self.directory[index] = Some(current_item)
+      self.real_size = self.real_size + 1
+      self.effective_size = self.effective_size + 1
+      index
+    }
+  }
+}
+
+fn index_of[T : Hash + Eq](self : MutableSet[T], key : T) -> Option[Int] {
+  let hash = key.hash().abs()
+  let num_slots = self.directory.length()
+  let slot = hash % num_slots
```

Comment:
hash may be negative
```moonbit
test "insert" {
  let set = MutableSet::[-1, 1, 2, 2, 3, 3]
  set.insert(4)
  @assertion.assert_eq(set.to_array(), [4, 3, 2, 1])?
}
```

Reply:
This error can be fixed using the same strategy as #93:
```
key.hash().abs().land(self.directory.length() - 1)
```

Although the `hash().abs()` calculation results of -1 and 1 are the same, they cannot be inserted at the same time, so the difference in `directory.length()` can avoid the same index.